README.md
=====================
yanexmoney-kit-django
=====================

Модуль оплаты yanexmoney-kit-django необходим для интеграции с сервисом
[Яндекс.Касса](http://kassa.yandex.ru/) на базе фраймворка
[Django](https://www.djangoproject.com).

Доступные платежные методы, если вы работаете как юридические лицо:
* **Банковские карты** — Visa (включая Electron), MasterCard и Maestro любого банка мира
* **Электронные деньги** — Яндекс.Деньги, WebMoney и QIWI Wallet
* **Наличные** — [Более 170 тысяч пунктов](https://money.yandex.ru/pay/doc.xml?id=526209) оплаты по России
* **Баланс телефона** — Билайн, МегаФон и МТС
* **Интернет банкинг** — Альфа-Клик, Сбербанк Онлайн, MasterPass и Промсвязьбанк
* **Кредитование**

### Минимальные требования:
* Django версии 2 или выше
* lxml версии 4.1.1

### Установка модуля
* Установить пакет:
```
pip install git+https://github.com/yandex-money/yandex-money-kit-django.git
```
* Добавить ``yandex_money`` в ``settings.INSTALLED_APPS``:
```
INSTALLED_APPS = (
    ...
    'yandex_money',
    ...
)
```
* Выполнить синхронизацию с БД:
```
python manage.py migrate
```
* Добавить в ``urls.py``:
```
urlpatterns = patterns('',
    # ...
    url(r'^fail-payment/$', TemplateView.as_view(template_name='fail.html'), name='payment_fail'),
    url(r'^success-payment/$', TemplateView.as_view(template_name='success.html'), name='payment_success'),
    url(r'^yandex-money/', include('yandex_money.urls')),
)
```
* Указать в settings.py следующие параметры:
```
YANDEX_MONEY_DEBUG = False
YANDEX_MONEY_SCID = 12345
YANDEX_MONEY_SHOP_ID = 56789
YANDEX_MONEY_SHOP_PASSWORD = 'password'
YANDEX_MONEY_FAIL_URL = 'https://example.com/fail-payment/'
YANDEX_MONEY_SUCCESS_URL = 'https://example.com/success-payment/'
# информировать о случаях, когда модуль вернул Яндекс.Кассе ошибку
YANDEX_MONEY_MAIL_ADMINS_ON_PAYMENT_ERROR = True
```
* Указать в личном кабинете Яндекс.Деньги настройки для приема уведомлений:

* paymentAvisoURL: `https://example.com/yandex-money/aviso/`
* checkURL: `https://example.com/yandex-money/check/`
* failURL: `https://example.com/fail-payment/`
* successURL: `https://example.com/success-payment/`


### Использование
[Полный пример использования](../example)

* Представление платежной формы:
```
from django.views.generic import TemplateView
from yandex_money.forms import PaymentForm
from yandex_money.models import Payment


class OrderPage(TemplateView):
    template_name = 'order_page.html'

    def get_context_data(self, **kwargs):
        payment = Payment(order_amount=123)
        payment.save()

        ctx = super(OrderPage, self).get_context_data(**kwargs)
        ctx['form'] = PaymentForm(instance=payment)
        return ctx
```
* Шаблон платежной формы:
```
<html>
    <head>
        <meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
        <meta charset="utf-8">
    </head>
    <body>
        <div style="border: 1px dotted gray; padding: 15px 15px 0; margin: 30px auto; width: 300px;">
            <form name="ShopForm" method="POST" action="https://money.yandex.ru/eshop.xml">
                <ul style="list-style: none;">
                    <li style="margin-bottom: 20px;">
                        Сумма заказа: <b>{{ form.sum.value }}</b>
                    </li>

                    {{ form.as_ul }}

                    <li style="margin-top: 20px;">
                        <input type="submit" value="Оплатить">
                    </li>
                </ul>
            </form>
        </div>
    </body>
</html>
```
* Также вы можете определить logger 'yandex_money':
```
LOGGING = {
    'version': 1,
    'disable_existing_loggers': True,
    'root': {
        'level': 'INFO',
        'handlers': ['default'],
    },
    'formatters': {
        'simple': {
            'format': '[%(asctime)s] %(levelname)s: %(message)s',
            'datefmt': '%d/%m/%Y %H:%M:%S',
        },
    },
    'handlers': {
        'default': {
            'level': 'DEBUG',
            'class': 'logging.handlers.RotatingFileHandler',
            'filename': os.path.join(LOGGING_DIR, 'messages.log'),
            'maxBytes': 1024 * 1024 * 5,
            'backupCount': 5,
            'formatter': 'message',
        },
        'yandex_money': {
            'level': 'DEBUG',
            'class': 'logging.handlers.RotatingFileHandler',
            'filename': os.path.join(LOGGING_DIR, 'yandex_money.log'),
            'maxBytes': 1024 * 1024 * 5,
            'backupCount': 5,
            'formatter': 'standard',
        },
    },
    'loggers': {
        '': {
            'handlers': ['default'],
            'level': 'INFO',
            'propagate': True
        },
        'yandex_money': {
            'handlers': ['yandex_money'],
            'level': 'DEBUG',
            'propagate': False
        },
    }
}
```

### Лицензионный договор.

Любое использование Вами программы означает полное и безоговорочное принятие Вами условий лицензионного договора, размещенного по адресу https://money.yandex.ru/doc.xml?id=527132 (далее – «Лицензионный договор»).
Если Вы не принимаете условия Лицензионного договора в полном объёме, Вы не имеете права использовать программу в каких-либо целях.

### Нашли ошибку или у вас есть предложение по улучшению модуля?
Пишите нам cms@yamoney.ru
При обращении необходимо:
* Указать наименование CMS и компонента магазина, а также их версии
* Указать версию платежного модуля (доступна в списке меню `Модули`)
* Описать проблему или предложение
* Приложить снимок экрана (для большей информативности)


CHANGELOG.md
### 82.0.0 от 27.02.2018 (форк)
* Удалена поддержка Python 2
* Добавлена поддержка Django 2+

### v1.1.3 от 02.02.2018
* Удален метод QP
* Добавлено кредитование

### 01.10.2015
* Добавлены новые платежные методы QW,QP

### 12.08.2015
* Внесены небольшие исправления

### 20.06.2015
* Добавлены новые платежные методы (MA, PB)
* Публикация лицензионного договора

### 15.05.2015
* Создан файл изменений


yandex_money/views.py
# -*- coding: utf-8 -*-

import logging

from django.http import HttpResponse
from django.utils import timezone
from django.utils.decorators import method_decorator
from django.views.decorators.csrf import csrf_exempt
from django.views.generic import View
from django.conf import settings
from django.core.mail import mail_admins
from lxml import etree
from lxml.builder import E

from .forms import CheckForm
from .forms import NoticeForm
from .models import Payment


logger = logging.getLogger('yandex_money')


class YandexValidationError(Exception):
    params = None

    def __init__(self, params=None):
        super(YandexValidationError, self).__init__()
        self.params = params if params is not None else {}


class BaseView(View):
    form_class = None

    @method_decorator(csrf_exempt)
    def dispatch(self, *args, **kwargs):
        return super(BaseView, self).dispatch(*args, **kwargs)

    def post(self, request, *args, **kwargs):
        form = self.form_class(request.POST)
        if form.is_valid():
            cd = form.cleaned_data
            if form.check_md5(cd):
                payment = self.get_payment(cd)
                if payment:
                    try:
                        self.validate(cd, payment)
                    except YandexValidationError as exc:
                        params = exc.params
                    else:
                        params = self.get_response_params(payment, cd)
                        self.mark_payment(payment, cd)
                        payment.send_signals()
                else:
                    params = {'code': '1000'}
            else:
                params = {'code': '1'}
        else:
            params = {'code': '200'}

        self.logging(request, params)
        content = self.get_xml(params)

        if (
            getattr(settings, 'YANDEX_MONEY_MAIL_ADMINS_ON_PAYMENT_ERROR', True) and
            params.get('code') != '0'
        ):
            mail_admins('yandexmoney_django error', 'post data: {post_data}\n\nresponse:{response}'.format(
                post_data=request.POST,
                response=content,
            ))

        return HttpResponse(content, content_type='application/xml')

    def validate(self, data, payment):
        pass

    def get_payment(self, cd):
        try:
            payment = Payment.objects.get(
                order_number=cd['orderNumber'], shop_id=cd['shopId'])
        except Payment.DoesNotExist:
            payment = None
        return payment

    def get_response_params(self, payment, cd):
        if payment:
            now = timezone.now()

            payment.performed_datetime = now
            payment.save()

            return {
                'code': '0',
                'shopId': str(cd['shopId']),
                'invoiceId': str(cd['invoiceId']),
                'performedDatetime': now.isoformat()
            }

        return {'code': '100'}

    def mark_payment(self, payment, cd):
        pass

    def get_xml(self, params):
        element = self.get_xml_element(**params)
        return etree.tostring(
            element,
            pretty_print = True,
            xml_declaration = True,
            encoding = 'UTF-8'
        )

    def get_xml_element(self, **params):
        raise NotImplementedError()

    def logging(self, request, params):
        message = 'Action %s has code %s for customerNumber "%s"' % (
            request.POST.get('action', ''), params['code'],
            request.POST.get('customerNumber', ''))
        logger.info(message)


class CheckOrderFormView(BaseView):
    form_class = CheckForm

    def validate(self, data, payment):
        if payment.order_amount != data['orderSumAmount']:
            params = {
                'code': '100',
                'message': 'Неверно указана сумма платежа',
            }
            raise YandexValidationError(params=params)

    def get_xml_element(self, **params):
        return E.checkOrderResponse(**params)


class NoticeFormView(BaseView):
    form_class = NoticeForm

    def get_xml_element(self, **params):
        return E.paymentAvisoResponse(**params)

    def mark_payment(self, payment, cd):
        payment.cps_email = cd.get('cps_email', '')
        payment.cps_phone = cd.get('cps_phone', '')
        payment.order_currency = cd.get('orderSumCurrencyPaycash')
        payment.shop_amount = cd.get('shopSumAmount')
        payment.shop_currency = cd.get('shopSumCurrencyPaycash')
        payment.payer_code = cd.get('paymentPayerCode', '')
        payment.payment_type = cd.get('paymentType')
        payment.status = payment.STATUS.SUCCESS
        payment.save()


yandex_money/models.py
# -*- coding: utf-8 -*-

from uuid import uuid4

import six
from django.conf import settings
from django.db import models

from .signals import payment_process
from .signals import payment_completed


def get_default_as_uuid():
    return six.text_type(uuid4()).replace('-', '')

class Payment(models.Model):
    class STATUS:
        PROCESSED = 'processed'
        SUCCESS = 'success'
        FAIL = 'fail'

        CHOICES = (
            (PROCESSED, 'Processed'),
            (SUCCESS, 'Success'),
            (FAIL, 'Fail'),
        )

    class PAYMENT_TYPE:
        PC = 'PC'
        AC = 'AC'
        GP = 'GP'
        MC = 'MC'
        WM = 'WM'
        SB = 'SB'
        AB = 'AB'
        MA = 'MA'
        PB = 'PB'
        QW = 'QW'
        CR = 'CR'

        CHOICES = (
            (PC, 'Кошелек Яндекс.Деньги'),
            (AC, 'Банковская карта'),
            (GP, 'Наличными через кассы и терминалы'),
            (MC, 'Счет мобильного телефона'),
            (WM, 'Кошелек WebMoney'),
            (SB, 'Сбербанк: оплата по SMS или Сбербанк Онлайн'),
            (AB, 'Альфа-Клик'),
            (MA, 'MasterPass'),
            (PB, 'Интернет-банк Промсвязьбанка'),
            (QW, 'QIWI Wallet'),
            (CR, 'Заплатить по частям'),
        )

    class CURRENCY:
        RUB = 643
        TEST = 10643

        CHOICES = (
            (RUB, 'Рубли'),
            (TEST, 'Тестовая валюта'),
        )

    user = models.ForeignKey(
        settings.AUTH_USER_MODEL,
        on_delete = models.PROTECT,
        blank = True,
        null = True,
        verbose_name = 'Пользователь'
    )

    pub_date = models.DateTimeField(
        'Время создания',
        auto_now_add = True
    )

    # Required request fields
    shop_id = models.PositiveIntegerField(
        'ID магазина',
        default = settings.YANDEX_MONEY_SHOP_ID
    )

    scid = models.PositiveIntegerField(
        'Номер витрины',
        default = settings.YANDEX_MONEY_SCID
    )

    customer_number = models.CharField(
        'Идентификатор плательщика',
        max_length = 64,
        default = get_default_as_uuid
    )

    order_amount = models.DecimalField(
        'Сумма заказа',
        max_digits = 15,
        decimal_places = 2
    )

    # Non-required fields
    article_id = models.PositiveIntegerField(
        'Идентификатор товара',
        blank = True,
        null = True
    )

    payment_type = models.CharField(
        'Способ платежа',
        max_length = 2,
        default = PAYMENT_TYPE.PC,
        choices = PAYMENT_TYPE.CHOICES
    )

    order_number = models.CharField(
        'Номер заказа',
        max_length = 64,
        default = get_default_as_uuid
    )

    cps_email = models.EmailField(
        'Email плательщика',
        max_length = 100,
        blank = True,
        null = True
    )

    cps_phone = models.CharField(
        'Телефон плательщика',
        max_length = 15,
        blank = True,
        null = True
    )

    success_url = models.URLField(
        'URL успешной оплаты',
        default = settings.YANDEX_MONEY_SUCCESS_URL
    )

    fail_url = models.URLField(
        'URL неуспешной оплаты',
        default = settings.YANDEX_MONEY_FAIL_URL
    )

    # Transaction info
    status = models.CharField(
        'Статус',
        max_length = 16,
        choices = STATUS.CHOICES,
        default = STATUS.PROCESSED
    )

    invoice_id = models.PositiveIntegerField(
        'Номер транзакции оператора',
        blank = True,
        null = True
    )

    shop_amount = models.DecimalField(
        'Сумма полученная на р/с',
        max_digits = 15,
        decimal_places = 2,
        blank = True,
        null = True,
        help_text = 'За вычетом процента оператора'
    )

    order_currency = models.PositiveIntegerField(
        'Валюта',
        default = CURRENCY.RUB,
        choices = CURRENCY.CHOICES
    )

    shop_currency = models.PositiveIntegerField(
        'Валюта полученная на р/с',
        blank = True,
        null = True,
        default = CURRENCY.RUB,
        choices = CURRENCY.CHOICES
    )

    performed_datetime = models.DateTimeField(
        'Время выполнение запроса',
        blank = True,
        null = True
    )

    @property
    def is_payed(self):
        return self.status == self.STATUS.SUCCESS

    def send_signals(self):
        status = self.status
        if status == self.STATUS.PROCESSED:
            payment_process.send(sender = self)
        if status == self.STATUS.SUCCESS:
            payment_completed.send(sender = self)

    @classmethod
    def get_used_shop_ids(cls):
        return cls.objects.values_list('shop_id', flat = True).distinct()

    @classmethod
    def get_used_scids(cls):
        return cls.objects.values_list('scid', flat = True).distinct()

    class Meta:
        ordering = ('-pub_date',)
        unique_together = (
            ('shop_id', 'order_number'),
        )
        verbose_name = 'платёж'
        verbose_name_plural = 'платежи'
        app_label = 'yandex_money'

    def __str__(self):
        return '[Payment id={}, order_number={}, payment_type={}, status={}]'.format(
            self.id, self.order_number, self.payment_type, self.status)


yandex_money/forms.py
# -*- coding: utf-8 -*-

from hashlib import md5

import six
from django import forms
from django.conf import settings

from .models import Payment


class BasePaymentForm(forms.Form):
    """
        shopArticleId               <no use>
        scid                        scid
        sum                         amount
        customerNumber              user
        orderNumber                 id
        shopSuccessURL              success_url
        shopFailURL                 fail_url
        cps_provider                payment_type
        cps_email                   cps_email
        cps_phone                   cps_phone
        paymentType                 payment_type
        shopId                      shop_id
        invoiceId                   invoice_id
        orderCreatedDatetime        <no use>
        orderSumAmount              order_amount
        orderSumCurrencyPaycash     order_currency
        orderSumBankPaycash         <no use>
        shopSumAmount               shop_amount
        shopSumCurrencyPaycash      shop_currency
        shopSumBankPaycash          <no use>
        paymentPayerCode            payer_code
        paymentDatetime             <no use>
        cms_name                    django
    """

    class ERROR_MESSAGE_CODES:
        BAD_SCID = 0
        BAD_SHOP_ID = 1

    error_messages = {
        ERROR_MESSAGE_CODES.BAD_SCID: 'scid не совпадает с YANDEX_MONEY_SCID',
        ERROR_MESSAGE_CODES.BAD_SHOP_ID: 'scid не совпадает с YANDEX_MONEY_SHOP_ID'
    }

    class ACTION:
        CHECK = 'checkOrder'
        CPAYMENT = 'paymentAviso'

        CHOICES = (
            (CHECK, 'Проверка заказа'),
            (CPAYMENT, 'Уведомления о переводе'),
        )

    shopId = forms.IntegerField(
        initial = settings.YANDEX_MONEY_SHOP_ID
    )

    scid = forms.IntegerField(
        initial = settings.YANDEX_MONEY_SCID
    )

    orderNumber = forms.CharField(
        min_length = 1,
        max_length = 64
    )

    customerNumber = forms.CharField(
        min_length = 1,
        max_length = 64
    )

    paymentType = forms.CharField(
        label = 'Способ оплаты',
        widget = forms.Select(
            choices = Payment.PAYMENT_TYPE.CHOICES
        ),
        min_length = 2,
        max_length = 2,
        initial = Payment.PAYMENT_TYPE.PC
    )

    orderSumBankPaycash = forms.IntegerField()

    md5 = forms.CharField(
        min_length = 32,
        max_length = 32
    )

    action = forms.CharField(
        max_length = 16
    )

    def __init__(self, *args, **kwargs):
        super(BasePaymentForm, self).__init__(*args, **kwargs)
        if hasattr(settings, 'YANDEX_ALLOWED_PAYMENT_TYPES'):
            allowed_payment_types = settings.YANDEX_ALLOWED_PAYMENT_TYPES
            self.fields['paymentType'].widget.choices = filter(
                lambda x: x[0] in allowed_payment_types,
                self.fields['paymentType'].widget.choices)

    @classmethod
    def make_md5(cls, cd):
        """
        action;orderSumAmount;orderSumCurrencyPaycash;orderSumBankPaycash;shopId;invoiceId;customerNumber;shopPassword
        """
        return md5(';'.join(map(six.text_type, (
            cd['action'],
            cd['orderSumAmount'],
            cd['orderSumCurrencyPaycash'],
            cd['orderSumBankPaycash'],
            cd['shopId'],
            cd['invoiceId'],
            cd['customerNumber'],
            settings.YANDEX_MONEY_SHOP_PASSWORD,
        ))).encode('utf-8')).hexdigest().upper()

    @classmethod
    def check_md5(cls, cd):
        return cls.make_md5(cd) == cd['md5']

    def clean_scid(self):
        scid = self.cleaned_data['scid']
        if (
            scid != settings.YANDEX_MONEY_SCID and
            not scid in Payment.get_used_scids()
        ):
            raise forms.ValidationError(self.error_messages[self.ERROR_MESSAGE_CODES.BAD_SCID])
        return scid

    def clean_shopId(self):
        shop_id = self.cleaned_data['shopId']
        if (
            shop_id != settings.YANDEX_MONEY_SHOP_ID and
            not shop_id in Payment.get_used_shop_ids()
        ):
            raise forms.ValidationError(self.error_messages[self.ERROR_MESSAGE_CODES.BAD_SHOP_ID])
        return shop_id


class PaymentForm(BasePaymentForm):
    sum = forms.FloatField(
        label = 'Сумма заказа'
    )

    cps_email = forms.EmailField(
        label = 'Email',
        required = False
    )

    cps_phone = forms.CharField(
        label = 'Телефон',
        max_length = 15,
        required = False
    )

    shopFailURL = forms.URLField(
        initial = settings.YANDEX_MONEY_FAIL_URL
    )

    shopSuccessURL = forms.URLField(
        initial = settings.YANDEX_MONEY_SUCCESS_URL
    )

    def __init__(self, *args, **kwargs):
        instance = kwargs.pop('instance')
        super(PaymentForm, self).__init__(*args, **kwargs)

        self.fields.pop('md5')
        self.fields.pop('action')
        self.fields.pop('orderSumBankPaycash')

        if not getattr(settings, 'YANDEX_MONEY_DEBUG', False):
            for name in self.fields:
                if name not in self.get_display_field_names():
                    self.fields[name].widget = forms.HiddenInput()

        if instance:
            self.fields['sum'].initial = instance.order_amount
            self.fields['paymentType'].initial = instance.payment_type
            self.fields['customerNumber'].initial = instance.customer_number
            self.fields['orderNumber'].initial = instance.order_number
            if instance.fail_url:
                self.fields['shopFailURL'].initial = instance.fail_url
            if instance.success_url:
                self.fields['shopSuccessURL'].initial = instance.success_url
            if instance.cps_email:
                self.fields['cps_email'].initial = instance.cps_email
            if instance.cps_phone:
                self.fields['cps_phone'].initial = instance.cps_phone

    def get_display_field_names(self):
        return ['paymentType', 'cps_email', 'cps_phone']


class CheckForm(BasePaymentForm):
    invoiceId = forms.IntegerField()

    orderSumAmount = forms.DecimalField(
        min_value = 0,
        decimal_places = 2
    )

    orderSumCurrencyPaycash = forms.IntegerField()

    shopSumAmount = forms.DecimalField(
        min_value = 0,
        decimal_places = 2
    )

    shopSumCurrencyPaycash = forms.IntegerField()

    paymentPayerCode = forms.IntegerField(
        min_value = 1,
        required = False
    )


class NoticeForm(BasePaymentForm):
    invoiceId = forms.IntegerField(
        min_value = 1
    )

    orderSumAmount = forms.DecimalField(
        min_value = 0,
        decimal_places = 2
    )

    orderSumCurrencyPaycash = forms.IntegerField()

    shopSumAmount = forms.DecimalField(
        min_value = 0,
        decimal_places = 2
    )

    shopSumCurrencyPaycash = forms.IntegerField()

    paymentPayerCode = forms.IntegerField(
        min_value = 1,
        required = False
    )

    cps_email = forms.EmailField(
        required = False
    )

    cps_phone = forms.CharField(
        max_length = 15,
        required = False
    )
